### PR TITLE
tpm2_eventlog: verifies digest matches event payload for certain event types

### PR DESCRIPTION
**Pull request by huang195**
This PR attempts to verify eventlog entries by comparing digest values against payload, to thwart possible forgeries.

Relevant issues: Issue #2252, Issue #2247 

As I addressed in Issue #2252, verifying all event types is currently impossible from the eventlog alone. There are certain event types, e.g., EV_EFI_BOOT_SERVICES_APPLICATION, EV_EFI_BOOT_SERVICES_DRIVER, that are used to load driver, kernel, shim, grub code into the memory during the boot process, where their digest is calculated based on the loaded code, which is not included in the eventlog. There needs to be an out-of-band process to verify these events along with the in-band eventlog digest verification that we're doing here. Moreover, ambiguities in the spec could lead to vendor specific way of determining what data is used to calculate the digest. Also bugs in vendor firmware could also contribute to more variations. 

So, as a result, I felt that we should not fail-stop when digest verification fails, but simply print out warning messages and continue to parse the rest of the eventlog. Afterwards, another entity could take the warning messages and the parsed eventlog to further determine how to handle digest verification failure. 

This PR is currently only handling a few event types that we see consistent behaviors across different vendors. Once we have more understanding of other event types, their verification can be gradually added.

### CODE REVIEW

**Comment by huang195** [comment]
@idesai @williamcroberts  any idea why Travis is failing on only 1 of the 12 jobs? I looked at the failed tests, and it doesn't look like the PR should have an effect on them, but then again, I might be wrong. Any suggestions?

**Comment by idesai** [comment]
> @idesai @williamcroberts any idea why Travis is failing on only 1 of the 12 jobs? I looked at the failed tests, and it doesn't look like the PR should have an effect on them, but then again, I might be wrong. Any suggestions?

I just had to re-run the specific build on travis.

File: lib/tpm2_eventlog.c [hunk 1]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+ * For event types where digest can be verified from their event payload,
+ * perform verification to ensure event payload was not tempered
+ */
+bool verify_digests(size_t eventnum, TCG_EVENT_HEADER2 const *eventhdr, TCG_EVENT2 *event) {
+    size_t i;
+
+    TCG_DIGEST2 const *digest = eventhdr->Digests;
+    UINT32 digest_count = eventhdr->DigestCount;
+    UINT32 event_type = eventhdr->EventType;
+
```
Comment (idesai) [inline lib/tpm2_eventlog.c:295]:
```suggestion
```

File: lib/tpm2_eventlog.c [hunk 2]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+    switch (event_type) {
+
```
Comment (idesai) [inline lib/tpm2_eventlog.c:297]:
```suggestion
```

File: lib/tpm2_eventlog.c [hunk 3]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+    /* Digests of these event types are calculated directly from event->Event, thus can be verified  */
+    case EV_S_CRTM_VERSION:
+    case EV_SEPARATOR:
+    case EV_EFI_VARIABLE_DRIVER_CONFIG:
+    case EV_EFI_GPT_EVENT:
+        for (i = 0; i < digest_count; i++) {
+            TPMI_ALG_HASH alg = digest->AlgorithmId;
+            size_t alg_size = tpm2_alg_util_get_hash_size(alg);
```
Comment (idesai) [inline lib/tpm2_eventlog.c:305]:
```suggestion
```

Comment (idesai) [reply lib/tpm2_eventlog.c:305]: Define closest to its first use

File: lib/tpm2_eventlog.c [hunk 4]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+            TPM2B_DIGEST calc_digest;
+
+            if (!tpm2_openssl_hash_compute_data(alg, event->Event, event->EventSize, &calc_digest)) {
+                LOG_WARN("Event %lu: Cannot calculate hash value from data", eventnum-1);
+                return false;
+            }
+            if (memcmp(calc_digest.buffer, digest->Digest, alg_size) != 0) {
```
Comment (idesai) [inline lib/tpm2_eventlog.c:312]:
```suggestion

            size_t alg_size = tpm2_alg_util_get_hash_size(alg);
            if (memcmp(calc_digest.buffer, digest->Digest, alg_size) != 0) {
```

File: lib/tpm2_eventlog.c [hunk 5]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+    case EV_S_CRTM_VERSION:
+    case EV_SEPARATOR:
+    case EV_EFI_VARIABLE_DRIVER_CONFIG:
+    case EV_EFI_GPT_EVENT:
+        for (i = 0; i < digest_count; i++) {
+            TPMI_ALG_HASH alg = digest->AlgorithmId;
+            size_t alg_size = tpm2_alg_util_get_hash_size(alg);
+            TPM2B_DIGEST calc_digest;
+
+            if (!tpm2_openssl_hash_compute_data(alg, event->Event, event->EventSize, &calc_digest)) {
```
Comment (idesai) [inline lib/tpm2_eventlog.c:308]:
```suggestion
            bool result = !tpm2_openssl_hash_compute_data(alg, event->Event,
            event->EventSize, &calc_digest);
            if (!result) {
```

Comment (idesai) [reply lib/tpm2_eventlog.c:308]: we don't hide a function return in a conditional statement.

File: lib/tpm2_eventlog.c [hunk 6]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                LOG_WARN("Event %lu: Cannot calculate hash value from data", eventnum-1);
+                return false;
+            }
+            if (memcmp(calc_digest.buffer, digest->Digest, alg_size) != 0) {
+                LOG_WARN("Event %lu's digest does not match its payload", eventnum-1);
```
Comment (idesai) [inline lib/tpm2_eventlog.c:313]:
```suggestion
                LOG_WARN("Event %lu's digest does not match its payload",
                eventnum - 1);
```

File: lib/tpm2_eventlog.c [hunk 7]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                return false;
+            }
+            digest = (TCG_DIGEST2*)((uintptr_t)digest->Digest + alg_size);
```
Comment (idesai) [inline lib/tpm2_eventlog.c:316]:
```suggestion

            digest = (TCG_DIGEST2*)((uintptr_t)digest->Digest + alg_size);
```

File: lib/tpm2_eventlog.c [hunk 8]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+        }
+        break;
+
+    /* Shim and grub use this event type for various tasks */
+    case EV_IPL:
+
```
Comment (idesai) [inline lib/tpm2_eventlog.c:322]:
```suggestion
```

File: lib/tpm2_eventlog.c [hunk 9]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+        /* PCR9: used to measure loaded kernel and initramfs images which cannot
+           be verified from eventlog alone */
+        if (eventhdr->PCRIndex == 9)
+            return true;
+
```
Comment (idesai) [inline lib/tpm2_eventlog.c:327]:
```suggestion
        if (eventhdr->PCRIndex == 9) {
            return true;
        }

```

File: lib/tpm2_eventlog.c [hunk 10]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+        /* PCR14: used to measure MokList, MokListX, and MokSBState which cannot
+           be verified from eventlog alone */
+        if (eventhdr->PCRIndex == 14)
+            return true;
+
```
Comment (idesai) [inline lib/tpm2_eventlog.c:332]:
```suggestion
        if (eventhdr->PCRIndex == 14) {
            return true;
        }

```

File: lib/tpm2_eventlog.c [hunk 11]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+            LOG_WARN("Event %lu is unexpectedly not extending either PCR 8, 9, or 14", eventnum-1);
+            return false;
+        }
+
+        /* Digest is applied on the string between "^[a-zA-Z-]+: " and "\0$" */
+        for (i = 0; i < digest_count; i++) {
+            size_t j;
+            TPMI_ALG_HASH alg = digest->AlgorithmId;
+            size_t alg_size = tpm2_alg_util_get_hash_size(alg);
+            TPM2B_DIGEST calc_digest;
```
Comment (idesai) [inline lib/tpm2_eventlog.c:344]:
```suggestion
```

Comment (idesai) [reply lib/tpm2_eventlog.c:344]: Define closest to first use

File: lib/tpm2_eventlog.c [hunk 12]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+
+            for (j = 0; j < event->EventSize; j++) {
+                if (event->Event[j] == ' ')
+                    break;
+            }
```
Comment (idesai) [inline lib/tpm2_eventlog.c:349]:
```suggestion
            }

```

File: lib/tpm2_eventlog.c [hunk 13]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+            if (j+1 >= event->EventSize || event->Event[event->EventSize-1] != '\0') {
```
Comment (idesai) [inline lib/tpm2_eventlog.c:350]:
```suggestion
            if (j + 1 >= event->EventSize || event->Event[event->EventSize - 1] != '\0') {
```

File: lib/tpm2_eventlog.c [hunk 14]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                LOG_WARN("Event %lu's event data is in unexpected format", eventnum-1);
```
Comment (idesai) [inline lib/tpm2_eventlog.c:351]:
```suggestion
                LOG_WARN("Event %lu's event data is in unexpected format", eventnum - 1);
```

File: lib/tpm2_eventlog.c [hunk 15]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                return false;
+            }
+
+            if (!tpm2_openssl_hash_compute_data(alg, event->Event+j+1, event->EventSize-j-2, &calc_digest)) {
```
Comment (idesai) [inline lib/tpm2_eventlog.c:355]:
```suggestion
            TPM2B_DIGEST calc_digest;
            TPMI_ALG_HASH alg = digest->AlgorithmId;
            bool result = tpm2_openssl_hash_compute_data(alg,
            event->Event + (j + 1), event->EventSize - (j - 2), &calc_digest);
            if (!result) {
```

File: lib/tpm2_eventlog.c [hunk 16]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                LOG_WARN("Event %lu: Cannot calculate hash value from data", eventnum-1);
```
Comment (idesai) [inline lib/tpm2_eventlog.c:356]:
```suggestion
                LOG_WARN("Event %lu: Cannot calculate hash value from data", eventnum - 1);
```

File: lib/tpm2_eventlog.c [hunk 17]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                return false;
+            }
+            if (memcmp(calc_digest.buffer, digest->Digest, alg_size) != 0) {
```
Comment (idesai) [inline lib/tpm2_eventlog.c:359]:
```suggestion

            size_t alg_size = tpm2_alg_util_get_hash_size(alg);
            if (memcmp(calc_digest.buffer, digest->Digest, alg_size) != 0) {
```

File: lib/tpm2_eventlog.c [hunk 18]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                LOG_WARN("Event %lu's digest does not match its payload", eventnum-1);
```
Comment (idesai) [inline lib/tpm2_eventlog.c:360]:
```suggestion
                LOG_WARN("Event %lu's digest does not match its payload", eventnum - 1);
```

File: lib/tpm2_eventlog.c [hunk 19]
```diff
@@ -281,6 +282,96 @@ bool foreach_sha1_log_event(tpm2_eventlog_context *ctx, TCG_EVENT const *eventhd
...
+                return false;
+            }
+            digest = (TCG_DIGEST2*)((uintptr_t)digest->Digest + alg_size);
+        }
+        break;
+
+    /* Verification succeeds by default for all other event types */
+    default:
+        return true;
```
Comment (idesai) [inline lib/tpm2_eventlog.c:369]:
```suggestion
```

File: lib/tpm2_eventlog.c [hunk 20]
```diff
@@ -328,6 +419,10 @@ bool foreach_event2(tpm2_eventlog_context *ctx, TCG_EVENT_HEADER2 const *eventhd
             return ret;
         }
 
+        /* digest verification */
+        if (ctx->data != 0)
+            verify_digests(*(size_t*)ctx->data, eventhdr, event);
+
```
Comment (idesai) [inline lib/tpm2_eventlog.c:425]:
```suggestion
        if (ctx->data != 0) {
            verify_digests(*(size_t*)ctx->data, eventhdr, event);
        }

```

File: lib/tpm2_eventlog.c [hunk 21]
```diff
@@ -1,5 +1,6 @@
 #include <inttypes.h>
 #include <stdlib.h>
+#include <string.h>
 #include <tss2/tss2_tpm2_types.h>
```
Comment (idesai) [inline lib/tpm2_eventlog.c:4]:
```suggestion

#include <tss2/tss2_tpm2_types.h>
```

**Review by idesai — CHANGES_REQUESTED** [review]

**Comment by idesai** [comment]
@huang195 if this is a new eventype being parsed,  I'd appreciate you add the eventlog binary to our test fixtures. Thank you for this PR.

**Comment by huang195** [comment]
@idesai all the event types this PR checks are already included in one or more of the existing test fixtures. BTW, is there a particular formatter you're using as I could run that through before submitting PRs in the future. Thanks!

**Comment by idesai** [comment]
> @idesai all the event types this PR checks are already included in one or more of the existing test fixtures. BTW, is there a particular formatter you're using as I could run that through before submitting PRs in the future. Thanks!

@williamcroberts has one for eclipse [here](https://github.com/tpm2-software/tpm2-tools/blob/5505dc9893406ba6297f3b1711616f5ee5687760/misc/formatters/tpm2.0-eclipse-cdt-formatter.xml)

**Comment by idesai** [comment]
@huang195 please rebase these to a single commit.

**Comment by huang195** [comment]
@idesai squashed to a single commit

**Review by idesai — APPROVED** [review]

**Comment by williamcroberts** [comment]
> > @idesai @williamcroberts any idea why Travis is failing on only 1 of the 12 jobs? I looked at the failed tests, and it doesn't look like the PR should have an effect on them, but then again, I might be wrong. Any suggestions?
> 
> I just had to re-run the specific build on travis.

Looks like:
```
spawn tss2 sign --keyPath=HS/SRK/mySignKey --padding=RSA_PSS --digest=/tmp/tpm2_test_bqnEzM/tss2_fapi.i9ewEy/digest.file --signature=/tmp/tpm2_test_bqnEzM/tss2_fapi.i9ewEy/signature.file --publicKey=/tmp/tpm2_test_bqnEzM/tss2_fapi.i9ewEy/public_key.file
ERROR:fapi:src/tss2-fapi/ifapi_policy_callbacks.c:907:search_policy() ErrorCode (0x0006002e) Policy not found. 
ERROR:fapi:src/tss2-fapi/ifapi_policy_callbacks.c:1191:ifapi_exec_auth_policy() ErrorCode (0x0006002e) Search policy 
ERROR:fapi:src/tss2-fapi/ifapi_policy_execute.c:614:execute_policy_authorize() Execute authorized policy. ErrorCode (0x0006002e) 
ERROR:fapi:src/tss2-fapi/ifapi_policy_execute.c:1526:execute_policy_element() ErrorCode (0x0006002e) Execute policy authorize 
ERROR:fapi:src/tss2-fapi/ifapi_policy_execute.c:1736:ifapi_policyeval_execute() Execute policy. ErrorCode (0x0006002e) 
ERROR:fapi:src/tss2-fapi/ifapi_policyutil_execute.c:320:ifapi_policyutil_execute() Execute policy. ErrorCode (0x0006002e) 
WARNING:esys:src/tss2-esys/api/Esys_FlushContext.c:234:Esys_FlushContext_Finish() Received TPM Error 
ERROR:esys:src/tss2-esys/api/Esys_FlushContext.c:89:Esys_FlushContext() Esys Finish ErrorCode (0x000001c4) 
ERROR:fapi:src/tss2-fapi/fapi_util.c:2042:ifapi_authorize_object() Execute policy. ErrorCode (0x0006002e) 
ERROR:fapi:src/tss2-fapi/fapi_util.c:2737:ifapi_key_sign() ErrorCode (0x0006002e) Authorize signature key. 
ERROR:fapi:src/tss2-fapi/api/Fapi_Sign.c:299:Fapi_Sign_Finish() Fapi sign. ErrorCode (0x0006002e) 
ERROR:esys:src/tss2-esys/esys_iutil.c:1101:esys_GetResourceObject() Error: Esys handle does not exist (70018). 
ERROR:esys:src/tss2-esys/api/Esys_FlushContext.c:138:Esys_FlushContext_Async() flushHandle unknown. ErrorCode (0x00070018) 
ERROR:esys:src/tss2-esys/api/Esys_FlushContext.c:66:Esys_FlushContext() Error in async function ErrorCode (0x00070018) 
ERROR:fapi:src/tss2-fapi/api/Fapi_Sign.c:130:Fapi_Sign() ErrorCode (0x0006002e) Key_Sign 
Fapi_Sign(0x6002E) - fapi:The policy is unknown
The system has not asked for branch selection
```

I don't know why Fapi Sign would start failing.... @AndreasFuchsSIT did something break in the TSS libraries?

**Comment by williamcroberts** [comment]
Is it related to https://github.com/tpm2-software/tpm2-tools/pull/2262?

**Comment by huang195** [comment]
@williamcroberts my other PR is also failing at exactly the same places as this one, but failures don't seem to be relevant to the PRs.

**Comment by cplappert** [comment]
The issue is addressed in #2268.

**Comment by williamcroberts** [comment]
@huang195 can you rebase on top of master and force push your branch, the CI should be working again.

**Comment by huang195** [comment]
@williamcroberts rebased, but I'm not sure what the error means. Can you please advise?

**Comment by huang195** [comment]
@williamcroberts rebased, but I'm not sure what the error means. Can you please advise?

**Comment by williamcroberts** [comment]
> @williamcroberts rebased, but I'm not sure what the error means. Can you please advise?

I just re-triggered cirrus, lets see if it resolves.

**Comment by huang195** [comment]
@williamcroberts yep, looks that did the trick

**Comment by williamcroberts** [comment]
Tested here:
https://github.com/williamcroberts/tpm2.0-tools/actions/runs/392273641

Merging.

**Pull request merged**